numpy/core/src/multiarray/item_selection.c

```diff
@@ -2128,6 +2129,69 @@ count_nonzero_bytes_384(const npy_uint64 * w)
     return r;
 }
 
+#if NPY_SIMD
+
+/* Count the zero bytes between `*d` and `end`, updating `*d` to point to where to keep counting from. */
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u8
+count_zero_bytes_u8(const npy_uint8 **d, const npy_uint8 *end)
+{
+    const npyv_u8 vone = npyv_setall_u8(1);
+    const npyv_u8 vzero = npyv_zero_u8();
+
+    npy_intp lane_max = 0;
+    npyv_u8 vsum8 = npyv_zero_u8();
+    while (*d < end && lane_max <= 0xFE) {
+        npyv_u8 vt = npyv_cvt_u8_b8(npyv_cmpeq_u8(npyv_load_u8(*d), vzero));
+        vt = npyv_and_u8(vt, vone);
+        vsum8 = npyv_add_u8(vsum8, vt);
+        *d += npyv_nlanes_u8;
+        lane_max += 1;
+    }
+    return vsum8;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u16
+count_zero_bytes_u16(const npy_uint8 **d, const npy_uint8 *end)
+{
+    npyv_u16 vsum16 = npyv_zero_u16();
+    npy_intp lane_max = 0;
+    while (*d < end && lane_max <= 0xFF00-0xFF) {
+        npyv_u8 vsum8 = count_zero_bytes_u8(d, end);
+        npyv_u16x2 part = npyv_expand_u16_u8(vsum8);
+        vsum16 = npyv_add_u16(vsum16, npyv_add_u16(part.val[0], part.val[1]));
+        lane_max += 0xFF*2;
+    }
+    return vsum16;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u32
+count_zero_bytes_u32(const npy_uint8 **d, const npy_uint8 *end)
+{
+    npyv_u32 vsum32 = npyv_zero_u32();
+    npy_intp lane_max = 0;
+    // The last accumulation needs to adjustment (2**32-1)/nlanes to avoid overflow.
+    while (*d < end && lane_max <= (0xFFFF0000-0xFFFF)/npyv_nlanes_u32) {
+        npyv_u16 vsum16 = count_zero_bytes_u16(d, end);
+        npyv_u32x2 part = npyv_expand_u32_u16(vsum16);
+        vsum32 = npyv_add_u32(vsum32, npyv_add_u32(part.val[0], part.val[1]));
+        lane_max += 0xFFFF*2;
+    }
+    return vsum32;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_intp
+count_nonzero_bytes(const npy_uint8 *d, npy_uintp unrollx)
+{
+    npy_intp zero_count = 0;
+    const npy_uint8 *end = d + unrollx;
+    while (d < end) {
+        npyv_u32 vsum32 = count_zero_bytes_u32(&d, end);
+        zero_count += npyv_sum_u32(vsum32);
+    }
+    return unrollx - zero_count;
+}
```

Comment:
One more idea:
```suggestion
/* Count the zero bytes between `*d` and `end`, updating `*d` to point to where to keep counting from. */
static NPY_INLINE NPY_GCC_OPT_3 npyv_u8
count_zero_bytes_u8(const npy_uint8 **d, const npy_uint8 *end, npy_uint8 max_count)
{
    const npyv_u8 vone = npyv_setall_u8(1);
    const npyv_u8 vzero = npyv_zero_u8();

    npy_intp lane_max = 0;
    npyv_u8 vsum8 = npyv_zero_u8();
    while (*d < end && lane_max <= max_count - 1) {
        npyv_u8 vt = npyv_cvt_u8_b8(npyv_cmpeq_u8(npyv_load_u8(*d), vzero));
        vt = npyv_and_u8(vt, vone);
        vsum8 = npyv_add_u8(vsum8, vt);
        *d += npyv_nlanes_u8;
        lane_max += 1;
    }
    return vsum8;
}

static NPY_INLINE NPY_GCC_OPT_3 npyv_u16
count_zero_bytes_u16(const npy_uint8 **d, const npy_uint8 *end, npy_uint16 max_count)
{
    npyv_u16 vsum16 = npyv_zero_u16();
    npy_intp lane_max = 0;
    while (*d < end && lane_max <= max_count - 0xFF*2) {
        npyv_u8 vsum8 = count_zero_bytes_u8(d, end, 0xFF);
        npyv_u16x2 part = npyv_expand_u16_u8(vsum8);
        vsum16 = npyv_add_u16(vsum16, npyv_add_u16(part.val[0], part.val[1]));
        lane_max += 0xFF*2;
    }
    return vsum16;
}

static NPY_INLINE NPY_GCC_OPT_3 npyv_u32
count_zero_bytes_u32(const npy_uint8 **d, const npy_uint8 *end, npy_uint32 max_count)
{
    npyv_u32 vsum32 = npyv_zero_u32();
    npy_intp lane_max = 0;
    while (*d < end && lane_max <= max_count - 2*0xFFFF) {
        npyv_u16 vsum16 = count_zero_bytes_u16(d, end, 0xFFFF);
        npyv_u32x2 part = npyv_expand_u32_u16(vsum16);
        vsum32 = npyv_add_u32(vsum32, npyv_add_u32(part.val[0], part.val[1]));
        lane_max += 0xFFFF*2;
    }
    return vsum32;
}

static NPY_INLINE NPY_GCC_OPT_3 npy_intp
count_nonzero_bytes(const npy_uint8 *d, npy_uintp unrollx)
{
    npy_intp zero_count = 0;
    const npy_uint8 *end = d + unrollx;
    while (d < end) {
        // The npyv_nlanes_u32 factor ensures that the sum of all lanes still fits in a uint32
        npyv_u32 vsum32 = count_zero_bytes_u32(&d, end, 0xFFFFFFFF / npyv_nlanes_u32);
        zero_count += npyv_sum_u32(vsum32);
    }
    return unrollx - zero_count;
}
```

Reply:
This is much more clear, one more suggestion, should we define three macro to identify the maximum value of uint8/16/32 because too many `0xFF` exists? such as MAX_UINT8/16/32.

Reply:
We already have those macros somewhere, yes - I think as `NPY_MAX_UINT8`?